File: src/runtime/bridge/state.ts
import { mkdirSync } from "node:fs";
import { readFile, rm, writeFile } from "node:fs/promises";
import { dirname } from "node:path";

export type BridgeRuntimeStatus = "starting" | "running" | "stopping" | "stopped" | "error";

export interface BridgeRuntimeState {
  readonly version: 1;
  readonly phase: "daemon";
  readonly status: BridgeRuntimeStatus;
  readonly updatedAt: string;
  readonly startedAt: string | null;
  readonly pid: number | null;
  readonly appName: string;
  readonly env: string;
  readonly logFilePath: string;
  readonly databaseFilePath: string | null;
  readonly activeRunCount: number;
  readonly activeSessionCount: number;
  readonly boundChatCount: number;
  readonly lastPollAt: string | null;
  readonly lastSuccessfulPollAt: string | null;
  readonly lastFailedPollAt: string | null;
  readonly consecutivePollFailures: number;
  readonly lastPollError: string | null;
  readonly previousOffset: number | null;
  readonly currentOffset: number | null;
  readonly lastEvent: string | null;
}

export async function readBridgeRuntimeState(filePath: string): Promise<BridgeRuntimeState | null> {
  try {
    const raw = await readFile(filePath, "utf8");
    const parsed = JSON.parse(raw) as Partial<BridgeRuntimeState>;
    if (!isBridgeRuntimeState(parsed)) {
      return null;
    }

    return parsed;
  } catch {
    return null;
  }
}

export async function writeBridgeRuntimeState(filePath: string, state: BridgeRuntimeState): Promise<void> {
  mkdirSync(dirname(filePath), { recursive: true });
  await writeFile(filePath, `${JSON.stringify(state, null, 2)}\n`, "utf8");
}

export async function removeBridgeRuntimeState(filePath: string): Promise<boolean> {
  try {
    await rm(filePath, { force: false });
    return true;
  } catch {
    return false;
  }
}

export async function readPidFile(filePath: string): Promise<number | null> {
  try {
    const raw = (await readFile(filePath, "utf8")).trim();
    if (!/^\d+$/.test(raw)) {
      return null;
    }

    return Number(raw);
  } catch {
    return null;
  }
}

export async function writePidFile(filePath: string, pid: number): Promise<void> {
  mkdirSync(dirname(filePath), { recursive: true });
  await writeFile(filePath, `${pid}\n`, "utf8");
}

export async function removePidFile(filePath: string): Promise<boolean> {
  try {
    await rm(filePath, { force: false });
    return true;
  } catch {
    return false;
  }
}

export function isProcessRunning(pid: number): boolean {
  try {
    process.kill(pid, 0);
    return true;
  } catch (error) {
    return !isProcessNotFoundError(error);
  }
}

function isBridgeRuntimeState(value: Partial<BridgeRuntimeState>): value is BridgeRuntimeState {
  return (
    value.version === 1 &&
    value.phase === "daemon" &&
    isRuntimeStatus(value.status) &&
    typeof value.updatedAt === "string" &&
    (value.startedAt === null || typeof value.startedAt === "string") &&
    (value.pid === null || typeof value.pid === "number") &&
    typeof value.appName === "string" &&
    typeof value.env === "string" &&
    typeof value.logFilePath === "string" &&
    (value.databaseFilePath === null || typeof value.databaseFilePath === "string") &&
    typeof value.activeRunCount === "number" &&
    typeof value.activeSessionCount === "number" &&
    typeof value.boundChatCount === "number" &&
    (value.lastPollAt === null || typeof value.lastPollAt === "string") &&
    (value.lastSuccessfulPollAt === null || typeof value.lastSuccessfulPollAt === "string") &&
    (value.lastFailedPollAt === null || typeof value.lastFailedPollAt === "string") &&
    typeof value.consecutivePollFailures === "number" &&
    (value.lastPollError === null || typeof value.lastPollError === "string") &&
    (value.previousOffset === null || typeof value.previousOffset === "number") &&
    (value.currentOffset === null || typeof value.currentOffset === "number") &&
    (value.lastEvent === null || typeof value.lastEvent === "string")
  );
}

function isRuntimeStatus(value: unknown): value is BridgeRuntimeStatus {
  return value === "starting" || value === "running" || value === "stopping" || value === "stopped" || value === "error";
}

function isProcessNotFoundError(error: unknown): error is NodeJS.ErrnoException {
  return Boolean(
    error &&
      typeof error === "object" &&
      "code" in error &&
      (error as NodeJS.ErrnoException).code === "ESRCH"
  );
}

File: internal/daemon/state.go
package daemon

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"syscall"
	"time"
)

type RuntimeStatus string

const (
	StatusStarting RuntimeStatus = "starting"
	StatusRunning  RuntimeStatus = "running"
	StatusStopping RuntimeStatus = "stopping"
	StatusStopped  RuntimeStatus = "stopped"
	StatusError    RuntimeStatus = "error"
)

type RuntimeState struct {
	Version              int           `json:"version"`
	Phase                string        `json:"phase"`
	Status               RuntimeStatus `json:"status"`
	UpdatedAt            string        `json:"updated_at"`
	StartedAt            string        `json:"started_at,omitempty"`
	PID                  *int          `json:"pid,omitempty"`
	LogFilePath          string        `json:"log_file_path"`
	DatabaseFilePath     string        `json:"database_file_path"`
	LastPollAt           string        `json:"last_poll_at,omitempty"`
	LastSuccessfulPollAt string        `json:"last_successful_poll_at,omitempty"`
	LastFailedPollAt     string        `json:"last_failed_poll_at,omitempty"`
	LastPollError        string        `json:"last_poll_error,omitempty"`
	PreviousOffset       *int64        `json:"previous_offset,omitempty"`
	CurrentOffset        *int64        `json:"current_offset,omitempty"`
	ActiveRunCount       int           `json:"active_run_count"`
	ActiveSessionCount   int           `json:"active_session_count"`
	LastEvent            string        `json:"last_event,omitempty"`
}

func ReadRuntimeState(path string) (*RuntimeState, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, nil
		}
		return nil, err
	}

	var state RuntimeState
	if err := json.Unmarshal(raw, &state); err != nil {
		return nil, nil
	}
	if state.Version != 1 || state.Phase != "daemon" {
		return nil, nil
	}
	return &state, nil
}

func WriteRuntimeState(path string, state RuntimeState) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o700); err != nil {
		return err
	}

	body, err := json.MarshalIndent(state, "", "  ")
	if err != nil {
		return err
	}
	body = append(body, '\n')

	tempFile, err := os.CreateTemp(filepath.Dir(path), ".bridge-state-*.json")
	if err != nil {
		return err
	}
	tempPath := tempFile.Name()
	if _, err := tempFile.Write(body); err != nil {
		tempFile.Close()
		_ = os.Remove(tempPath)
		return err
	}
	if err := tempFile.Close(); err != nil {
		_ = os.Remove(tempPath)
		return err
	}
	if err := os.Chmod(tempPath, 0o600); err != nil {
		_ = os.Remove(tempPath)
		return err
	}
	return os.Rename(tempPath, path)
}

func ReadPIDFile(path string) (int, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return 0, nil
		}
		return 0, err
	}
	value, err := strconv.Atoi(strings.TrimSpace(string(raw)))
	if err != nil || value <= 0 {
		return 0, nil
	}
	return value, nil
}

func WritePIDFile(path string, pid int) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o700); err != nil {
		return err
	}
	return os.WriteFile(path, []byte(strconv.Itoa(pid)+"\n"), 0o600)
}

func RemovePIDFile(path string) error {
	err := os.Remove(path)
	if errors.Is(err, os.ErrNotExist) {
		return nil
	}
	return err
}

func IsProcessRunning(pid int) bool {
	if pid <= 0 {
		return false
	}
	process, err := os.FindProcess(pid)
	if err != nil {
		return false
	}
	return process.Signal(syscall.Signal(0)) == nil
}

func TailFile(path string, lineCount int) ([]string, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	lines := strings.Split(strings.ReplaceAll(string(raw), "\r\n", "\n"), "\n")
	filtered := make([]string, 0, len(lines))
	for _, line := range lines {
		if line != "" {
			filtered = append(filtered, line)
		}
	}
	if lineCount >= len(filtered) {
		return filtered, nil
	}
	return filtered[len(filtered)-lineCount:], nil
}

func nowRFC3339() string {
	return time.Now().UTC().Format(time.RFC3339)
}
